### DISCUSSION TITLE
How to create a component story with a custom demo, controls enabled, and basic component code via Show Code?

### DISCUSSION THREAD

**Discussion by roypearce** [storybookjs/storybook#29201] [Help] [answered]
### Summary

Has anybody figured out how to write a story that will use custom code for the visual demo, the demo is configurable with the arg controls, and when `Show Code` is pressed, will reveal code specific to the component implementation with the customized args instead of the code used for the visual demo?

Using storybook/react 7.6.

In this example code, I am customizing the visual demo of a button for simplicity, however, my use case is for more complex components like dialogs which require a user interaction in order to display them, that would require a button in the visual demo to launch the dialog, but the desired output of the `Show Code` would be specific to the dialog component, not how the story was setup for the demo.

```
const meta: Meta<typeof Button> = {
  title: 'Components/Button',
  component: Button,
  argTypes: {
    layout: {
      control: 'select',
      options: ['default', 'secondary', 'outline', 'text'],
    },
    size: {
      control: 'select',
      options: ['small', 'default', 'large'],
    },
    disabled: { control: 'boolean' },
  },
};

export default meta;
type Story = StoryObj<typeof Button>;

export const Standard: Story = {
  args: {
    layout: 'default',
    size: 'default',
    disabled: false,
    children: 'Click me',
  },
  render: (args) => (
    <div style={{ padding: '20px', background: '#f0f0f0' }}>
      <Button {...args} onClick={() => console.log('Button clicked!')} />
    </div>
  ),
  parameters: {
    docs: {
      source: {
        code: (context) => `
<Button
  layout="${context.args.layout}"
  size="${context.args.size}"
  disabled={${context.args.disabled}}
  onClick={() => console.log('Button clicked!')}
>
  ${context.args.children}
</Button>
`,
      },
    },
  },
};
```

The above example code will get me a custom demo w/ the button, but when I click on `Show Code` I get the following output:
```
context => `
<Button
  layout="${context.args.layout}"
  size="${context.args.size}"
  disabled={${context.args.disabled}}
  onClick={() => console.log('Button clicked!')}
>
  ${context.args.children}
</Button>
`
```

The desired output would input the control values from the story UI and be ready to be copied and inserted into app code:
```
<Button
  layout="outline"
  size="large"
  disabled={false}
  onClick={() => console.log('Button clicked!')}
>
  Click me
</Button>
```

I've tried several different configurations and options to no avail. I'm hoping someone from the community does stuff similar to this and can offer a working suggestion. Any help is appreciated. Thanks.

**Comment by mattbtay** [comment]
I'm fairly new to SB, but this is the solution I've gone with:
```
parameters: {
      docs: {
        source: {
          code: '
            <Button
               layout="outline"
               size="large"
               disabled={false}
               onClick={() => console.log('Button clicked!')}
            >
              Click me
            </Button>'
         }
     }
```


you can add any string representation of the component you want to visualize something other than the actual component. I've done this because in the story i have a few decorators that wrap my component, I want to show one of those decorators in the actual implementation as an example.

**Reply by roypearce** [reply]
> I'm fairly new to SB, but this is the solution I've gone with:
> 
> ```
> parameters: {
>       docs: {
>         source: {
>           code: '
>             <Button
>                layout="outline"
>                size="large"
>                disabled={false}
>                onClick={() => console.log('Button clicked!')}
>             >
>               Click me
>             </Button>'
>          }
>      }
> ```
> 
> you can add any string representation of the component you want to visualize something other than the actual component. I've done this because in the story i have a few decorators that wrap my component, I want to show one of those decorators in the actual implementation as an example.

Thanks for that @mattbtay you can indeed display custom code, but that is still limited in its utility because it is static.

When you use the string representation for the source code, it doesn't allow a dev to use the arg controls in the UI to update the source code so that they can configure a component in the UI and copy the resulting code for use in an application.

**Reply by roypearce** [reply, marked as answer]
I've managed to come up with a solution that ticks all the boxes (the args panel is populated with the component types defined in typescript): 
```
const args = {
  disabled: false,
  layout: "outline",
  onClick: () => console.log("Button clicked!"),
  size: "large"
};

const meta = {
  component: Button,
  title: 'Components/Button',
} satisfies Meta<typeof Button>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Standard: Story = {
  args,
  render: (args) => (
    <section>
      <Button {...args}>Click Me</Button>
    </section>
  ),
  parameters: {
    docs: {
      source: {
        transform: (_code, storyContext) => {
          const { args } = storyContext;
          const argString = Object.entries(args)
            .map(([key, value]) => {
              if (typeof value === 'function') {
                return `${key}={${value.toString()}}`;
              }
              return `${key}={${JSON.stringify(value)}}`;
            })
            .join('\n    ');
          return `<Button\n    ${argString}\n/>`;
        },
      },
    },
  },
};
```

This story will yield customizable args that update the generated code with their updated values as they are modified, and update the component demo as well. This is the resulting generated code from `Show Code`, note the `<section>` code is gone that only existed for this arbitrary exampl:

```
<Button
  disabled={false}
  layout="outline"
  onClick={() => console.log("Button clicked!")}
  size="large"
>
  Click Me
</Button>
```

The piece that I was missing was the `transform` function which stringifies the current args from the args panel as they are modified.
